File: kakao-blind-mock-test/kakao-71.js
MAX_INT = 1e100

function solution(strs, t) {
    const setStrs = new Set(strs);
	const dp = new Map();
    dp.set(0, 0)
    for (let j = 1; j <= t.length; j++) {
        dp.set(j, MAX_INT)
        for (let i = 1; i <= 5; i++) {
            if (j-i < 0) {
                continue;
            }
            if (setStrs.has(t.substring(j-i, j))) {
                dp.set(j, Math.min(dp.get(j), dp.get(j-i)+1))
            }
        }
    }
    const ret = dp.get(t.length)
    
	if (ret === MAX_INT) {
		return -1
    }
	return ret;
}

// console.log(solution(["banan", "n", "a"],"banana")===2)
// console.log(solution(["banan", "n", "a"],"banana")===2)
console.log(solution(["ba", "na", "n", "a"],"banana")===3)
// console.log(solution(["ba", "na", "n", "b"],"b")===1)
// console.log(solution(["app","ap","p","l","e","ple","pp"],"apple")===2)
// console.log(solution(["ba","an","nan","ban","n"],"banana")=== -1)

File: kakao-blind-mock-test/kakao-71.py
from collections import deque
def solution(strs, t):
    length = len(t)
    dp = dict()
    dp[''] = 0
    q = deque([''])
    while q:
        cur = q.popleft()
        for each_str in strs:
            if cur+each_str in dp:
                dp[cur+each_str] = min(dp[cur+each_str], dp[cur]+1)
            else:
                dp[cur+each_str] = dp[cur]+1
                if len(cur) + len(each_str) < length:
                    q.append(cur+each_str)
    if t in dp:
        return dp[t]
    else:
        return -1

ex_strs = ["ba", "na", "n", "a"]
ex_t = "banana"
print(solution(ex_strs,ex_t))
